src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -217,6 +229,67 @@ public void ConnectHeaders_SupplyHeadersThenSupplyMore_Both_Tip_PeerId_Maps_Shou
+        {
+            var testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+            testContext.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+
+            // PEER 1: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // PEER 2: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 7);
+            chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // PEER 3: 1a - 2a - 3a - 4a - 5b - 6b - 7b - 8b
+            BlockHeader forkedBlockHeader = testContext.GetNewBlock(chainTip.Previous.Previous.Previous.Previous).Header;
+            chainedHeaderTree.ConnectNewHeaders(3, new List<BlockHeader>() { forkedBlockHeader });
+            ChainedHeader peerThreeChainedHeader = chainedHeaderTree.GetChainedHeaderByPeerId(3);
+            ChainedHeader peerThreeTip = testContext.ExtendAChain(3, peerThreeChainedHeader);
+            List<BlockHeader> listOfNewAndExistingHeaders = testContext.ChainedHeaderToList(peerThreeTip, 4);
+            chainedHeaderTree.ConnectNewHeaders(3, listOfNewAndExistingHeaders );
+
+            // CHT should contain 12 headers
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 12);
+             
+            Dictionary<uint256, ChainedHeader> chainedHeadersByHash = chainedHeaderTree.GetChainedHeadersByHash();
+            List<ChainedHeader> chainedHeadersByHashList = chainedHeadersByHash.Values.ToList();
+
+            var blocksAtEachHeight = from chainedHeader in chainedHeadersByHashList
+                group chainedHeader by chainedHeader.Height
+                into heightGroup
+                select heightGroup;
+            
+            foreach (var blockGroup in blocksAtEachHeight)
+            {
+                // 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+                //                   - 5b - 6b - 7b - 8b
+                var height = blockGroup.Key;
+                if (height < 4) Assert.True(blockGroup.Count() == 1);
+                if (height > 4) Assert.True(blockGroup.Count() == 2);
+
+                // Each should have 1 Next pointer
+                if (height < 3 || (height > 3 && height < 7)) Assert.True(blockGroup.All(x => x.Next.Count == 1));
+
+                // Except for 8a and 8b which contain none
+                if (height == 7) Assert.True(blockGroup.All(x => x.Next.Count == 0));
+
+                // And 4a which contains 2
+                if (height == 3) Assert.True(blockGroup.All(x => x.Next.Count == 2));
+            }
```

Comment:
It seems correct, but the whole construction is unnecessarily complex. Why don't we just start at the tip of one chain and go back using previous, then take the tip of the other chain and do the same. We already have the fork point in a variable.

Reply:
Changed for this approach - now reverse iteration from chain tips